src/libmeasurement_kit/ooni/ooni_test.cpp

```diff
@@ -46,17 +46,19 @@ void OoniTest::run_next_measurement(Callback<Error> cb) {
         logger->debug("net_test: tearing down");
         teardown(next_input);
 
-        Error error = file_report.write_entry(entry);
-        if (error) {
-            cb(error);
-            return;
-        }
-        if (entry_cb) {
-            entry_cb(entry.dump());
-        }
-        logger->debug("net_test: written entry");
+        report.fill_entry(entry);
+        report.write_entry(entry, [=](Error error) {
+            if (error) {
+                cb(error);
+                return;
+            }
+            if (entry_cb) {
```

Comment:
I think it's better to call the `entry_cb` function before you handle the error callback. Actually it may even make sense to call it before you write the entry itself. I would imagine a user of this feature may want to know the result of the measurement independently from wether or not reporting succeeded.